src/hpke.rs

```diff
@@ -0,0 +1,1046 @@
+    fn response_state(&mut self) -> Result<AeadContext, error::Unspecified> {
+        let mut key = [0u8; MAX_AEAD_KEY_LEN];
+        self.export(
+            "response key".as_bytes(),
+            self.aead_algorithm.key_len() as u16,
+            &mut key[..self.aead_algorithm.key_len()],
+        )?;
+        let mut base_nonce = [0u8; aead::NONCE_LEN];
+        self.export(
+            "response nonce".as_bytes(),
+            self.aead_algorithm.nonce_len() as u16,
+            &mut base_nonce,
+        )?;
+        Ok(AeadContext {
+            sealing_key: make_key(
+                self.aead_algorithm,
+                &key[..self.aead_algorithm.key_len()],
+                &HpkeNonce {
+                    base_nonce,
+                    nonce_counter: [0u8; aead::NONCE_LEN],
+                },
+            ),
+            opening_key: make_key(
+                self.aead_algorithm,
+                &key[..self.aead_algorithm.key_len()],
+                &HpkeNonce {
+                    base_nonce,
+                    nonce_counter: [0u8; aead::NONCE_LEN],
+                },
+            ),
+            raw_key: key,
+            base_nonce,
+        })
+    }
+
+    fn export(
+        &self,
+        exporter_context: &[u8],
+        length: u16,
+        exported: &mut [u8],
+    ) -> Result<(), error::Unspecified> {
+        labeled_expand(
+            &self.suite_id_context,
+            self.hash,
+            &self.exporter_secret[..self.hash.len()],
+            "sec",
+            exporter_context,
+            length,
+            exported,
+        )
+    }
+}
+
+#[derive(Debug)]
+struct AeadContext {
+    sealing_key: aead::SealingKey<HpkeNonce>,
+    opening_key: aead::OpeningKey<HpkeNonce>,
+    raw_key: [u8; MAX_AEAD_KEY_LEN],
+    base_nonce: [u8; 12],
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct HpkeNonce {
+    base_nonce: [u8; aead::NONCE_LEN],
+    nonce_counter: [u8; aead::NONCE_LEN],
+}
+
+/// Increments the bytes, assuming the most significant
+/// bit is first, and returns the carry.
+fn increment_bytes(buf: &mut [u8], mut amount: u64) {
+    let mut i = buf.len() - 1;
+
+    while amount > 0 {
+        amount += buf[i] as u64;
+        buf[i] = amount as u8;
+        amount /= 256;
+
+        if i == 0 {
+            break;
+        }
+        i -= 1;
+    }
+}
+
+impl aead::NonceSequence for HpkeNonce {
+    fn advance(&mut self) -> Result<aead::Nonce, error::Unspecified> {
+        let mut new_nonce = [0u8; aead::NONCE_LEN];
+        for (i, (&x1, &x2)) in self
+            .base_nonce
+            .iter()
+            .zip(self.nonce_counter.iter())
+            .enumerate()
+        {
+            new_nonce[i] = x1 ^ x2;
+        }
+        increment_bytes(&mut self.nonce_counter, 1);
+        aead::Nonce::try_assume_unique_for_key(&new_nonce)
+    }
+}
+
+impl Suite {
+    /// new
+    pub fn new(kem: Kem, kdf_id: KdfId, aead_id: AeadId) -> Result<Self, error::Unspecified> {
```

Comment:
if we remove Kem structure, then
```suggestion
    pub fn new(kem_id: KemId, kdf_id: KdfId, aead_id: AeadId) -> Result<Self, error::Unspecified> {
```